pkg/operator/ceph/object/spec.go

```diff
@@ -912,3 +922,63 @@ func renderProbe(cfg rgwProbeConfig) (string, error) {
 
 	return writer.String(), nil
 }
+
+func (c *clusterConfig) addDNSNamesToRGWServer() (string, error) {
+	if (c.store.Spec.Hosting == nil) || len(c.store.Spec.Hosting.DNSNames) <= 0 {
+		return "", nil
+	}
+	if !c.clusterInfo.CephVersion.IsAtLeastReef() {
+		return "", errors.New("rgw dns names are supported from ceph v18 onwards")
+	}
+
+	// add default RGW service name to dns names
+	dnsNames := c.store.Spec.Hosting.DNSNames
+	dnsNames = append(dnsNames, domainNameOfService(c.store))
+
+	// add custom endpoints from zone spec if exists
+	if c.store.Spec.Zone.Name != "" {
+		zone, err := c.context.RookClientset.CephV1().CephObjectZones(c.store.Namespace).Get(c.clusterInfo.Context, c.store.Spec.Zone.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		for _, endpoint := range zone.Spec.CustomEndpoints {
+			url, err := url.Parse(endpoint)
+			if err != nil {
+				return "", err
+			}
+			domainName := strings.TrimPrefix(url.Hostname(), "www.")
+			if domainName != "" {
+				dnsNames = append(dnsNames, domainName)
+			}
```

Comment:
To feel certain the implementation here, let's make this a helper function, and make sure to thoroughly test the helper function in unit tests.

Reply:
is `zone.Spec.CustomEndpoints` specifically for use in `rgw_dns_name`, or does it do other things too? i'm not sure why a user would put www.something if they didn't want the www. part

Reply:
`zone.Spec.CustomEndpoints` is the list of zone endpoints that are used for RGW multisite replication traffic. They are not user-facing endpoints, but my understanding is that those endpoints do still need to be in the `rgw_dns_name` list based on conversations from the last month or so.

Reply:
if an endpoint says `http://www.example.com`, we should set `rgw_dns_name=www.example.com`. vhost-style requests would look like `bucketname.www.example.com`